lib/active_merchant/billing/gateways/square.rb

```diff
@@ -0,0 +1,375 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class SquareGateway < Gateway
+
+      # test uses a 'sandbox' access token, same URL
+      self.live_url = 'https://connect.squareup.com/v2/'
+
+      self.money_format = :cents
+      self.supported_countries = ['US', 'CA']
+      self.default_currency = 'USD'
+      self.supported_cardtypes = [:visa, :master, :american_express, :discover, :diners_club, :jcb, :maestro]
+
+      self.homepage_url = 'https://squareup.com/developers'
+      self.display_name = 'Square'
+
+      # Map to Square's https://docs.connect.squareup.com/api/connect/v2/#handlingerrors
+      STANDARD_ERROR_CODE_MAPPING = {
+        'INVALID_CARD' => STANDARD_ERROR_CODE[:invalid_number],
+        'INVALID_EXPIRATION' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'INVALID_EXPIRATION_DATE' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+        'INVALID_EXPIRATION_YEAR' => STANDARD_ERROR_CODE[:invalid_expiry_date],
+
+        # Something invalid in the card, e.g. verify declined when linking card to customer.
+        'INVALID_CARD_DATA' => STANDARD_ERROR_CODE[:processing_error],
+        'CARD_EXPIRED' => STANDARD_ERROR_CODE[:expired_card],
+        'VERIFY_CVV_FAILURE' => STANDARD_ERROR_CODE[:incorrect_cvc],
+        'VERIFY_AVS_FAILURE' => STANDARD_ERROR_CODE[:incorrect_zip],
+        'CARD_DECLINED' => STANDARD_ERROR_CODE[:card_declined],
+        'UNAUTHORIZED' => STANDARD_ERROR_CODE[:config_error]
+      }
+
+      # The `login` key is the client_id (also known as application id) in the dev portal
+      # Get it after you create a new app: https://connect.squareup.com/apps/
+      # The `password` is the access token (personal or OAuth)
+      # The `location_id` must be fetched initially https://docs.connect.squareup.com/articles/processing-payment-rest/
+      # The `test` indicates if these credentials are for sandbox or production (money moving) access
+      def initialize(options={})
+        requires!(options, :login, :password, :location_id, :test)
```

Comment:
activesupport also supplies
```ruby
options.assert_valid_keys(:login, :password, :location_id, :test)
```
which throws ArgumentError, in the event that the ruby `Hash#fetch` throwing KeyError isn't desirable